[Spam Prevention] Implements skeleton of application-layer feedback (#4099)
onflow/flow-go: network/alsp/readme.md

# Application Layer Spam Prevention (ALSP)
## Overview
The Application Layer Spam Prevention (ALSP) is a module that provides a mechanism to prevent the malicious nodes from 

> spamming the Flow nodes at the application layer (i.e., the engines). ASLP is not a multi-party protocol, i.e., 

Reviewer: ```suggestion
spamming the Flow nodes at the application layer (i.e., the engines). ALSP is not a multi-party protocol, i.e., 
```

it does not require the nodes to exchange any messages with each other for the purpose of spam prevention. Rather, it is 

> a local mechanism that is implemented by each node to protect itself from malicious nodes. AlSP is not meant to replace 

Reviewer 2: ```suggestion
a local mechanism that is implemented by each node to protect itself from malicious nodes. ALSP is not meant to replace 
```

the existing spam prevention mechanisms at the network layer (e.g., the Libp2p and GossipSub). 
Rather, it is meant to complement the existing mechanisms by providing an additional layer of protection.
ALSP is concerned with the spamming of the application layer through messages that appear valid to the networking layer and hence
are not filtered out by the existing mechanisms.

ALSP relies on the application layer to detect and report the misbehaviors that 
lead to spamming. It enforces a penalty system to penalize the misbehaving nodes that are reported by the application layer. ALSP also takes 

> extra measures to protect the network from malicious nodes that attempt on an active spamming attacks. Once the penalty of a remote node

Reviewer: ```suggestion
extra measures to protect the network from malicious nodes that attempt an active spamming attack. Once the penalty of a remote node
```

reaches a certain threshold, the local node will disconnect from the remote node and no-longer accept any incoming connections from the remote node 
until the penalty is reduced to zero again through a decaying interval.

## Features
- Spam prevention at the application layer. 
- Penalizes misbehaving nodes based on their behavior. 
- Configurable penalty values and decay intervals. 
- Misbehavior reports with customizable penalty amplification. 
- Thread-safe and non-blocking implementation. 

> - Maintains the safety and liveness of the Flow blockchain system by disallow-listing malicious nodes (i.e., application layer spammers).

Reviewer 2: ```suggestion
- Maintains the safety and liveness of the Flow blockchain system by block-listing malicious nodes (i.e., application layer spammers).
```

I think block-list is generally used as the opposite of allow-list. Might be worth changing the terminology to block-list, here and elsewhere in this file.

Author: Thank you for your suggestion! However, we have consciously chosen to adopt the "allow-listing" and 
"disallow-listing" terminology for the networking layer. This choice has been integrated throughout our entire networking layer codebase for consistency and clarity.

We find that using "block-list" can sometimes be confusing, as it might be mistaken for a module that manages a list of memory blocks. By using allow-listing and disallow-listing, we are able to avoid such confusion and ensure that our intentions are clearly communicated in our code.

## Architectural Principles
- **Non-intrusive**: ALSP is a local mechanism that is implemented by each node to protect itself from malicious nodes. It is not a multi-party protocol, i.e., it does not require the nodes to exchange any messages with each other for the purpose of spam prevention.
- **Non-blocking**: ALSP is non-blocking and does not affect the performance of the networking layer. It is implemented in a way that does not require the networking layer to wait for the ALSP to complete its operations. Non-blocking behavior is mandatory for the networking layer to maintain its performance. 
- **Thread-safe**: ALSP is thread-safe and can be used concurrently by multiple threads, e.g., concurrent engine calls on reporting misbehaviors.

## Usage

> ALSP is enabled by default through the networking layer. It is not necessary to explicitly enable it. Once can disable it by setting the `alsp-enbale` flag to `false`.

Reviewer 2: ```suggestion
ALSP is enabled by default through the networking layer. It is not necessary to explicitly enable it. One can disable it by setting the `alsp-enable` flag to `false`.
```
